src/analyses/call_graph.cpp

```diff
@@ -51,50 +46,111 @@ void call_grapht::add(
   const irep_idt &caller,
   const irep_idt &callee)
 {
-  graph.insert(std::pair<irep_idt, irep_idt>(caller, callee));
+  std::size_t caller_idx = node_numbering.number(caller);
+  if(caller_idx >= nodes.size())
+  {
+    node_indext node_index = add_node();
+    nodes[node_index].function_name = caller;
+  }
+
+  std::size_t callee_idx = node_numbering.number(callee);
+  if(callee_idx >= nodes.size())
+  {
+    node_indext node_index = add_node();
+    nodes[node_index].function_name = callee;
+  }
+
+  add_edge(caller_idx, callee_idx);
 }
 
-/// Returns an inverted copy of this call graph
-/// \return Inverted (callee -> caller) call graph
-call_grapht call_grapht::get_inverted() const
+
+void call_grapht::output_dot_node(std::ostream &out, node_indext n) const
 {
-  call_grapht result;
-  for(const auto &caller_callee : graph)
-    result.add(caller_callee.second, caller_callee.first);
-  return result;
+  const nodet &node = nodes.at(n);
+
+  for(const auto &edge : node.out)
+  {
+    out << "  \"" << node.function_name << "\" -> " << "\""
+        << nodes[edge.first].function_name << "\" " << " [arrowhead=\"vee\"];"
+        << "\n";
+  }
 }
 
-void call_grapht::output_dot(std::ostream &out) const
+void call_grapht::output_xml_node(std::ostream &out, node_indext n) const
 {
-  out << "digraph call_graph {\n";
+  const nodet &node = nodes.at(n);
 
-  for(const auto &edge : graph)
+  for(const auto &edge : node.out)
   {
-    out << "  \"" << edge.first << "\" -> "
-        << "\"" << edge.second << "\" "
-        << " [arrowhead=\"vee\"];"
-        << "\n";
+    out << "<call_graph_edge caller=\"";
+    xmlt::escape_attribute(id2string(node.function_name), out);
+    out << "\" callee=\"";
+    xmlt::escape_attribute(id2string(nodes[edge.first].function_name), out);
+    out << "\">\n";
   }
+}
 
-  out << "}\n";
+void call_grapht::output_xml(std::ostream &out) const
+{
+  for(node_indext n = 0; n < nodes.size(); n++)
```

Comment:
The reason for this for loop is that the output_xml_node function requires a node indext to be passed to it. 

The reason the output_xml_node function requires a node indext to be passed to it, rather than a reference to the node is to preserve the compatability with the output functions of grapht (see grapht<N>::output_dot_node). 

And the reason grapht<N>output_dot_node needs to have the node index passed to it is because it outputs that node index (instead of outputting data stored in the node, because the template doesn't know what might be contained in the node).

In order to use a C++11 for loop I would have to reimplement the output functions of grapht<T>, would you still prefer that?

Reply:
OK never mind then